net: Fix inet_pton IPv4 implementation

jukkar opened the pull request:
The conversion from IPv4 string presentation to numeric value did not check if the individual address value was between 0 and 255 inclusive.

Add also test case for this.

Fixes #84593

jukkar committed:
net: Fix inet_pton IPv4 implementation

The conversion from IPv4 string presentation to numeric value
did not check if the individual address value was between 0 and 255
inclusive.

Add also test case for this.

Fixes #84593

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on subsys/net/ip/utils.c:
```diff
@@ -349,8 +349,14 @@ int z_impl_net_addr_pton(sa_family_t family, const char *src,
 
-			addr->s4_addr[i] = strtol(src, &endptr, 10);
+			val = strtol(src, &endptr, 10);
+			if (val < 0 || val > 255) {
```
```suggestion
			if (val < 0 || val > UINT8_MAX) {
```

pdgendt commented on tests/net/socket/misc/src/main.c:
```diff
@@ -51,6 +51,9 @@ ZTEST_USER(socket_misc_test_suite, test_inet_pton)
 	res = zsock_inet_pton(AF_INET6, "a:b:c:d:0:1:2:3z", buf);
 	zassert_equal(res, 0, "");
+
+	res = zsock_inet_pton(AF_INET, "192.0.2.400", buf);
+	zassert_equal(res, 0, "");
```
Reading the implementation, I think you can easily read beyond the data. If the data is provided from an external source, and for example `"0.0"` is passed, this will call `strtol` to the data after the string (twice).

cfriedt commented:
I've always been curious as to why these `zsock_inet_pton()` and `zsock_inet_ntop()` were implemented as `__syscall`s. There are no kernel or private resources to synchronize on.

On machines with an mmu or mpu, there would be a fault of reading or writing outside of the allowed addressable memory.

In any case, the only thing I can suggest is to look at some of the degenerate cases such as the one that @pdgendt highlighted. But also, it would be good to cross-reference against e.g. the glibc or musl implementations or picolibc. See what kinds of invalid input those detect and try to match common behaviour.

Ofc, one could technically implement a bunch of exhaustive tests in the POSIX test suite, but it would be nice if the native Zephyr call were just compliant.

jukkar commented:
Added more checks (from stackoverflow) and added tests.
Just wondering should we disallow perfectly valid IPv4 addresses like `0`, `1`, `0.0` etc. (so values < 256). Atm we do not support the "plain int" addresses like `2130706433` which is `127.0.0.1`

pdgendt commented on tests/net/socket/misc/src/main.c:
```diff
@@ -51,6 +51,24 @@ ZTEST_USER(socket_misc_test_suite, test_inet_pton)
+	res = zsock_inet_pton(AF_INET, "1", buf);
+	zassert_equal(res, 1, "");
+
+	res = zsock_inet_pton(AF_INET, "256", buf);
+	zassert_equal(res, 0, "");
```
Is this valid?

cfriedt replied:
Iirc this function returns 1 if a valid address is matched, so I think the test seems to be checking that it's invalid (which looks right to me). Not sure if my intuition agrees with other implementations ATM tho

pdgendt commented on subsys/net/ip/utils.c:
```diff
@@ -349,8 +349,49 @@ int z_impl_net_addr_pton(sa_family_t family, const char *src,
 
 		for (i = 0; i < sizeof(struct in_addr); i++) {
```
The function currently loops all characters 3 times, an alternative with a single loop:

```c
		struct in_addr *addr = (struct in_addr *)dst;
		uint8_t index = 0, digits = 0;
		uint16_t value = 0;

		(void)memset(addr, 0, sizeof(struct in_addr));

		while (index < sizeof(struct in_addr)) {
			if (*src == '\0' || *src == '.') {
				if (digits == 0 || value > UINT8_MAX) {
					return -EINVAL;
				}

				addr->s4_addr[index] = value;

				if (*src == '\0') {
					break;
				}

				index++;
				digits = 0;
				value = 0;
			} else if ('0' <= *src && *src <= '9') {
				if (++digits > 3) {
					/* Number too large */
					return -EINVAL;
				}
				value *= 10;
				value += *src - '0';
			} else {
				/* Invalid character */
				return -EINVAL;
			}

			src++;
		}
```

cfriedt replied:
I asked chatgpt to generate a list of invalid IPv4 addresses. I think most things should be covered above, but a couple that stood out:

- empty segments (e.g. 192..1.0 - above, we would need to check that digits is also > 0)
- preceding zeros or non-singular zero (e.g. 192.168.000.1 or 192.168.0.01 - if the first digit of an octet is zero, the octet must be of length 1)

pdgendt replied:
Empty segments is checked with `digits == 0`. I thought about preceding zeroes, but didn't wanted to add the complexity..

Can be checked with the following code if the delimiter or the end was found.
```c
if ((digits > 1 && value < 10) || (digits > 2 && value < 100) {
	/* Preceding zeroes */
	return -EINVAL;
}
```

rlubos commented on tests/net/socket/misc/src/main.c:
```diff
@@ -51,6 +51,24 @@ ZTEST_USER(socket_misc_test_suite, test_inet_pton)
+	res = zsock_inet_pton(AF_INET, "192.0.2.400", buf);
+	zassert_equal(res, 0, "");
+
+	res = zsock_inet_pton(AF_INET, "0.0", buf);
```
I might have missed some discussion, but I'm not sure I follow why should these be considered a valid IP addresses? Even checked with native Linux app, and they're all rejected by `inet_pton()`:
```
inet_pton returned 0 for 0.0
inet_pton returned 0 for 0.1
inet_pton returned 0 for 0
inet_pton returned 0 for 1
```

cfriedt replied:
I think they probably are not valid IP addresses.

jukkar replied:
Ok, lets ban these short addresses then. I have seen these used as IP addresses but it probably was not with `inet_pton` but with some other tool.

piotrParf replied:
I do not think they are invalid at all, they may be unusable. Those adresses may be invalid on the public internet or have a designated not "free for all" usage.
The IANA list: https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml for example lists addresses 0.0.0.0 and its subnet as valid but to limited use.

jukkar commented:
* Updated according to comments

pdgendt commented on subsys/net/ip/utils.c:
```diff
@@ -335,24 +335,63 @@ int z_impl_net_addr_pton(sa_family_t family, const char *src,
-			addr->s4_addr[i] = strtol(src, &endptr, 10);
+		while (index < sizeof(struct in_addr)) {
+			if (*src == '\0' || *src == '.') {
+				if (*src == '.') {
+					count++;
+				}
```
You could move the `index++` here and verify that `index == sizeof(struct in_addr)` after the loop, that way you can remove `count` altogether.
